src/charm.py

```diff
@@ -110,6 +188,47 @@ def __init__(self, *args):
         self.framework.observe(
             self.on.validate_configuration_action, self._on_validate_configuration
         )
+        self.framework.observe(self.resources_patch.on.patch_failed, self._on_k8s_patch_failed)
+
+    @classmethod
+    def _limits_to_requests(cls, resource_limits: ResourceSpecDict):
+        """Calculate "requests" from a "limits" dict.
+
+        By default, we set the "requests" portion of the resource limits to a sensible value, while
+        keeping the "limits" portion unspecified.
+        If the user specifies a "limits" (via config options), then the "requests" portion is
+        calculated to maintain a monotone function, as follows:
```

Comment:
The use of phrase "monotone function" is strictly not correct here (please see my comments above). Would it make more sense to call this methods something like `_constrained_requests(limits: dict)` ? Thereby suggesting that this method constrains the kubernetes requests to honor the user specified limits.

Reply:
Fixed in next commit, lemme know what you think!